michbl/assignment3/Dictionary.hpp

```diff
@@ -0,0 +1,128 @@
+#ifndef DICTIONARY_HPP
+#define DICTIONARY_HPP
+
+#include <iostream>
+#include <vector>
+#include <string>
+#include <unordered_map>
+#include <set>
+#include "Grid.hpp"
+
+#define END_OF_WORD -1
+
+using namespace std;
+
+// implementing dictionary as trie
+class Dictionary {
+private:
+    struct Node {
+        char c;
+        unordered_map<char, Node *> children;
+
+        Node(char c) : c(c) {
+            this->children.clear();
+        }
+    };
+
+    Node *root;
+    vector<Node *> nodes; // for memory release
+
+    void wordSearchBT(Grid &grid, int row, int col, set<string> &res, string word) {
+        char c;
+        if (!grid.getChar(row, col, c)) {
+            return;
+        }
+        word += c;

```

Comment:
Tip:
You can pass `word` as a reference (or pointer) to this function, to avoid copying it again and again every function call (can be very inefficient if this code runs for very long words).
Upside: more efficient!
Downside: This way you need to remember to remove every letter after you added it, before returning from the function. (less rubust and less readable).

```C++
word.push_back(c);  // Adds c on the same buffer without allocating a new one.
if (isPrefix()) {
   for (Direction dir : dirs) { /* recursive call*/ }
}
word.pop();
return;
```

Reply:
The downside is why I wrote it like that. Do you think the upside wins in this case?

Reply:
Hmm, it really depends.
If we follow "Premature optimization is the root of all evil", then readability should be our first instinct. I am personally in favor of "functional" programming, making each function independent and share the least info between them.
In this case, this is a recursive call which can happen many times, and copying the string can have a real effect of the program runtime; but even this depends on the input (how long the words are, how large is the board...).

I think that most important is that you understand the implications of both ways.
If an interviewer will ask you what is the complexity of this solution, then you will be able to spot the cost of copying the word; and when they ask "can you do better", they you will say: "Yes, we can share one accumulating word between the functions".